libs/base/System/File.idr

```diff
@@ -288,26 +307,61 @@ fPoll (FHandle f)
     = do p <- primIO (prim__fPoll f)
          pure (p > 0)
 
+||| Produce enough @Fuel@ to read the given number
+||| of lines.
 export
-readFile : HasIO io => String -> io (Either FileError String)
-readFile file
+lineCount : Nat -> Fuel
+lineCount 0 = Dry
+lineCount (S k) = More (lineCount k)
+
+||| Read a chunk of a file in a line-delimited fashion.
+||| You can use this function to read an entire file
+||| as with @readFile@ by reading until @forever@ or by
+||| iterating through pages until hitting the end of
+||| the file.
+|||
+||| The @lineCount@ function can provide you with enough
+||| fuel to read exactly a given number of lines.
+|||
+||| On success, returns a tuple of whether the end of
+||| the file was reached or not and the String read in
+||| from the file.
+|||
+||| Note that because we are chunking by lines, this
+||| function's totality depends on the assumption that
+||| no single line in the input file is infinite.
+export
+readFilePage : HasIO io => (offset : Nat) -> (until : Fuel) -> String -> io (Either FileError (Bool, String))
```

Comment:
Would it make sense to get a `List String` instead?

I see that you are using `fastAppend` but that's still a lot of
appending especially if we are primarily interested in the lines? 

I guess we should at least expose the auxiliary function corresponding
to `read offset fuel []` plus the `openFile` error handling.

Reply:
Originally I thought it made sense to leave the return type closer to `readFile`, but I do like your idea of `List String`. 

One awkward thing is that each line still has a newline at the end but this is actually not an unprecedented behavior; I happen to know that in Elixir the `File.stream!/3` function defaults to reading by line with the same behavior (you need to drop the newline from the end of each line if you want to be rid of it).

My intuition for this function would be to return the equivalent of the file after running `lines` on it, but given the precedent I mentioned for the behavior of leaving the newlines in and the fact that doing so means fewer overall changes in this one PR, I vote we leave it this way (that is, change the signature to returning `List String` but accept that newlines are not stripped).

Reply:
Can you elaborate on the following?

> I guess we should at least expose the auxiliary function corresponding
> to read offset fuel [] plus the openFile error handling.

I thought you might mean roughly what I did in my latest changes where I pulled `withFile` and `readOnto` out but I also got a bit lost in the weeds trying to deduplicate code and might have made an ill-advised decision with my `try` function.